rfd/0212-jsonpath-interpolation.md

```diff
@@ -0,0 +1,134 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD 211 - JSONPath Interpolation
+
+## What
+
+Introduce a new interpolation function, `jsonpath`, which will use the
+[JSONPath query language](https://www.rfc-editor.org/rfc/rfc9535.html)
+to transform a json blob into a single value or list of values.
+
+Initially, `jsonpath` will be supported for:
+
+- [Role templates](https://goteleport.com/docs/admin-guides/access-controls/guides/role-templates/#interpolation-rules),
+  where the user trait value being evaluated is a json blob.
+- OIDC `claims_to_roles` mapping, where the claim value being evaluated is a json blob.
+
+## Why
+
+In some cases, OIDC claims may not be a string or list of strings (string value)
+as is currently expected by Teleport, making these claims unusable for both OIDC
+role mapping and role templates. `jsonpath` interpolation is primarily aimed at
+solving this use case.
+
+There may also be some local administrative use cases where `jsonpath`
+interpolation is useful, especially given some of the more powerful features
+of JSONPath parsing like filtering.
+
+## Details
+
+### Traits
+
+Traits will maintain the same key/value pair structure, where the value is either
+a string or list of strings. If the string value is a JSON blob, then it can be
+interpolated in role templates with `jsonpath`.
+
+For example:
+
+```yaml
+kind: user
+metadata:
+  name: alice
+spec:
+  ...
+  traits:
+    server_perms: '{
```

Comment:
why convert traits to string is this to avoid the complexity of implementation? what's the downside of storing it as embedded json?

Reply:
Embedded json should work, I suppose it's only on the gRPC layer that will have to flatten the JSON traits into a JSON blob.